File: typescript/src/reader.ts
import { BufferUnderflowError, InvalidWireTypeError, DecodeError } from "./errors";
import { WireType, TypeID, FieldTag, decodeTag, zigzagDecode, zigzagDecode64 } from "./types";

// Module-level singleton to avoid repeated instantiation
const textDecoder = new TextDecoder();

/**
 * Reader decodes Cramberry data from a binary buffer.
 */
export class Reader {
  private buffer: Uint8Array;
  private view: DataView;
  private pos: number;
  private end: number;

  constructor(data: Uint8Array) {
    this.buffer = data;
    this.view = new DataView(data.buffer, data.byteOffset, data.byteLength);
    this.pos = 0;
    this.end = data.length;
  }

  /**
   * Returns the current position in the buffer.
   */
  get position(): number {
    return this.pos;
  }

  /**
   * Returns the number of bytes remaining.
   */
  get remaining(): number {
    return this.end - this.pos;
  }

  /**
   * Returns true if there is more data to read.
   */
  get hasMore(): boolean {
    return this.pos < this.end;
  }

  /**
   * Checks if there are enough bytes available.
   */
  private checkAvailable(needed: number): void {
    if (this.pos + needed > this.end) {
      throw new BufferUnderflowError(needed, this.remaining);
    }
  }

  /**
   * Reads a raw byte.
   */
  readByte(): number {
    this.checkAvailable(1);
    return this.buffer[this.pos++];
  }

  /**
   * Reads raw bytes.
   */
  readBytes(length: number): Uint8Array {
    this.checkAvailable(length);
    const bytes = this.buffer.subarray(this.pos, this.pos + length);
    this.pos += length;
    return bytes;
  }

  /**
   * Maximum number of bytes for a varint (64-bit value encoded as varint).
   * A uint64 has 64 bits, and each varint byte encodes 7 bits,
   * so we need ceil(64/7) = 10 bytes maximum.
   */
  private static readonly MAX_VARINT_BYTES = 10;

  /**
   * Reads an unsigned varint (LEB128).
   * For 32-bit values, this uses the same 10-byte limit as 64-bit for consistency,
   * but the result is capped to 32 bits.
   */
  readVarint(): number {
    let result = 0;
    let shift = 0;

    for (let i = 0; i < Reader.MAX_VARINT_BYTES; i++) {
      this.checkAvailable(1);
      const b = this.buffer[this.pos++];

      // At the 5th byte (index 4), we've consumed 28 bits.
      // The 5th byte can only contribute 4 more bits for a 32-bit value.
      if (i === 4 && (b & 0xf0) !== 0) {
        throw new DecodeError("Varint overflow: value exceeds 32 bits");
      }

      result |= (b & 0x7f) << shift;
      if ((b & 0x80) === 0) {
        return result >>> 0; // Ensure unsigned
      }
      shift += 7;
    }

    throw new DecodeError("Varint overflow: exceeded 10 bytes");
  }

  /**
   * Reads an unsigned 64-bit varint (LEB128).
   * Uses a maximum of 10 bytes, consistent with protobuf and Go implementation.
   */
  readVarint64(): bigint {
    let result = 0n;
    let shift = 0n;

    for (let i = 0; i < Reader.MAX_VARINT_BYTES; i++) {
      this.checkAvailable(1);
      const b = this.buffer[this.pos++];
      const bBigInt = BigInt(b);

      // At the 10th byte (index 9), we've consumed 63 bits.
      // The 10th byte can only contribute 1 more bit (bit 63 of uint64).
      if (i === 9) {
        // If continuation bit is set, we'd need 11+ bytes
        if (b >= 0x80) {
          throw new DecodeError("Varint64 overflow: exceeded 10 bytes");
        }
        // If data portion is > 1, value would overflow uint64
        if (b > 1) {
          throw new DecodeError("Varint64 overflow: 10th byte must be 0 or 1");
        }
      }

      result |= (bBigInt & 0x7fn) << shift;
      if ((b & 0x80) === 0) {
        return result;
      }
      shift += 7n;
    }

    throw new DecodeError("Varint64 overflow: exceeded 10 bytes");
  }

  /**
   * Reads a signed varint using ZigZag decoding.
   */
  readSVarint(): number {
    return zigzagDecode(this.readVarint());
  }

  /**
   * Reads a signed 64-bit varint using ZigZag decoding.
   */
  readSVarint64(): bigint {
    return zigzagDecode64(this.readVarint64());
  }

  /**
   * Reads a field tag.
   */
  readTag(): FieldTag {
    const tag = this.readVarint();
    return decodeTag(tag);
  }

  /**
   * Reads a boolean.
   */
  readBool(): boolean {
    return this.readByte() !== 0;
  }

  /**
   * Reads a 32-bit signed integer.
   */
  readInt32(): number {
    return this.readSVarint();
  }

  /**
   * Reads a 64-bit signed integer as bigint.
   * Use this for full 64-bit precision.
   */
  readInt64(): bigint {
    return this.readSVarint64();
  }

  /**
   * Reads a 64-bit signed integer as a JavaScript number.
   *
   * WARNING: JavaScript numbers can only safely represent integers
   * up to Number.MAX_SAFE_INTEGER (2^53-1). Values larger than this
   * will lose precision.
   *
   * For values that may exceed 2^53-1, use readInt64() instead which
   * returns a bigint with full 64-bit precision.
   *
   * @param warnOnPrecisionLoss - If true (default), logs a warning when
   *                              precision loss occurs
   */
  readInt64AsNumber(warnOnPrecisionLoss: boolean = true): number {
    const value = this.readSVarint64();
    if (warnOnPrecisionLoss) {
      if (value > BigInt(Number.MAX_SAFE_INTEGER) ||
          value < BigInt(Number.MIN_SAFE_INTEGER)) {
        console.warn(
          `cramberry: int64 value ${value} exceeds safe integer range ` +
          `(${Number.MIN_SAFE_INTEGER} to ${Number.MAX_SAFE_INTEGER}), ` +
          `precision may be lost. Use readInt64() for full precision.`
        );
      }
    }
    return Number(value);
  }

  /**
   * Reads a 32-bit unsigned integer.
   */
  readUint32(): number {
    return this.readVarint();
  }

  /**
   * Reads a 64-bit unsigned integer as bigint.
   * Use this for full 64-bit precision.
   */
  readUint64(): bigint {
    return this.readVarint64();
  }

  /**
   * Reads a 64-bit unsigned integer as a JavaScript number.
   *
   * WARNING: JavaScript numbers can only safely represent integers
   * up to Number.MAX_SAFE_INTEGER (2^53-1). Values larger than this
   * will lose precision.
   *
   * For values that may exceed 2^53-1, use readUint64() instead which
   * returns a bigint with full 64-bit precision.
   *
   * @param warnOnPrecisionLoss - If true (default), logs a warning when
   *                              precision loss occurs
   */
  readUint64AsNumber(warnOnPrecisionLoss: boolean = true): number {
    const value = this.readVarint64();
    if (warnOnPrecisionLoss) {
      if (value > BigInt(Number.MAX_SAFE_INTEGER)) {
        console.warn(
          `cramberry: uint64 value ${value} exceeds safe integer range ` +
          `(max ${Number.MAX_SAFE_INTEGER}), precision may be lost. ` +
          `Use readUint64() for full precision.`
        );
      }
    }
    return Number(value);
  }

  /**
   * Reads a 32-bit float (IEEE 754).
   */
  readFloat32(): number {
    this.checkAvailable(4);
    const value = this.view.getFloat32(this.pos, true); // Little-endian
    this.pos += 4;
    return value;
  }

  /**
   * Reads a 64-bit float (IEEE 754).
   */
  readFloat64(): number {
    this.checkAvailable(8);
    const value = this.view.getFloat64(this.pos, true); // Little-endian
    this.pos += 8;
    return value;
  }

  /**
   * Reads a fixed 32-bit value.
   */
  readFixed32(): number {
    this.checkAvailable(4);
    const value = this.view.getUint32(this.pos, true); // Little-endian
    this.pos += 4;
    return value;
  }

  /**
   * Reads a fixed 64-bit value.
   */
  readFixed64(): bigint {
    this.checkAvailable(8);
    const value = this.view.getBigUint64(this.pos, true); // Little-endian
    this.pos += 8;
    return value;
  }

  /**
   * Reads a length-prefixed string.
   */
  readString(): string {
    const length = this.readVarint();
    const bytes = this.readBytes(length);
    return textDecoder.decode(bytes);
  }

  /**
   * Reads length-prefixed bytes.
   */
  readLengthPrefixedBytes(): Uint8Array {
    const length = this.readVarint();
    return this.readBytes(length);
  }

  /**
   * Skips a field based on its wire type.
   */
  skipField(wireType: WireType): void {
    switch (wireType) {
      case WireType.Varint:
      case WireType.SVarint:
        this.readVarint();
        break;
      case WireType.Fixed64:
        this.checkAvailable(8);
        this.pos += 8;
        break;
      case WireType.Bytes:
        const length = this.readVarint();
        this.checkAvailable(length);
        this.pos += length;
        break;
      case WireType.Fixed32:
        this.checkAvailable(4);
        this.pos += 4;
        break;
      case WireType.TypeRef:
        this.readVarint(); // Type ID
        // Then skip the actual value (as Bytes)
        const valueLength = this.readVarint();
        this.checkAvailable(valueLength);
        this.pos += valueLength;
        break;
      default:
        throw new InvalidWireTypeError(-1, wireType);
    }
  }

  /**
   * Creates a sub-reader for reading nested messages.
   */
  subReader(length: number): Reader {
    this.checkAvailable(length);
    const sub = new Reader(this.buffer.subarray(this.pos, this.pos + length));
    this.pos += length;
    return sub;
  }

  /**
   * Reads a type reference (for polymorphic types).
   * Returns the type ID and a sub-reader for the value data.
   * Format: [type_id: varint] [data_length: varint] [data: bytes]
   */
  readTypeRef(): { typeId: TypeID; reader: Reader } {
    const typeId = this.readVarint();
    const length = this.readVarint();
    const reader = this.subReader(length);
    return { typeId, reader };
  }
}

File: rust/src/reader.rs
//! Cramberry decoder.

use crate::error::{Error, Result};
use crate::types::{
    decode_compact_tag, zigzag_decode_32, zigzag_decode_64, FieldTag, WireType, END_MARKER,
};

/// Reader decodes Cramberry data from a binary buffer.
pub struct Reader<'a> {
    buffer: &'a [u8],
    pos: usize,
}

/// Maximum number of bytes for a varint-encoded uint64.
/// A uint64 has 64 bits, and each varint byte encodes 7 bits,
/// so we need ceil(64/7) = 10 bytes maximum.
const MAX_VARINT_BYTES: usize = 10;

impl<'a> Reader<'a> {
    /// Creates a new reader from a byte slice.
    pub fn new(data: &'a [u8]) -> Self {
        Self {
            buffer: data,
            pos: 0,
        }
    }

    /// Returns the current position in the buffer.
    pub fn position(&self) -> usize {
        self.pos
    }

    /// Returns the number of bytes remaining.
    pub fn remaining(&self) -> usize {
        self.buffer.len() - self.pos
    }

    /// Returns true if there is more data to read.
    pub fn has_more(&self) -> bool {
        self.pos < self.buffer.len()
    }

    /// Checks if there are enough bytes available.
    fn check_available(&self, needed: usize) -> Result<()> {
        if self.pos + needed > self.buffer.len() {
            return Err(Error::buffer_underflow(needed, self.remaining()));
        }
        Ok(())
    }

    /// Reads a raw byte.
    pub fn read_byte(&mut self) -> Result<u8> {
        self.check_available(1)?;
        let value = self.buffer[self.pos];
        self.pos += 1;
        Ok(value)
    }

    /// Reads raw bytes.
    pub fn read_bytes(&mut self, length: usize) -> Result<&'a [u8]> {
        self.check_available(length)?;
        let bytes = &self.buffer[self.pos..self.pos + length];
        self.pos += length;
        Ok(bytes)
    }

    /// Reads an unsigned varint (LEB128).
    /// For 32-bit values, this uses the same 10-byte limit as 64-bit for consistency,
    /// but the result is capped to 32 bits.
    pub fn read_varint(&mut self) -> Result<u32> {
        let mut result: u32 = 0;
        let mut shift = 0;

        for i in 0..MAX_VARINT_BYTES {
            self.check_available(1)?;
            let b = self.buffer[self.pos];
            self.pos += 1;

            // At the 5th byte (index 4), we've consumed 28 bits.
            // The 5th byte can only contribute 4 more bits for a 32-bit value.
            if i == 4 && (b & 0xf0) != 0 {
                return Err(Error::VarintOverflow);
            }

            result |= ((b & 0x7f) as u32) << shift;
            if b & 0x80 == 0 {
                return Ok(result);
            }
            shift += 7;
        }

        Err(Error::VarintOverflow)
    }

    /// Reads an unsigned 64-bit varint (LEB128).
    /// Uses a maximum of 10 bytes, consistent with protobuf and Go implementation.
    pub fn read_varint64(&mut self) -> Result<u64> {
        let mut result: u64 = 0;
        let mut shift = 0;

        for i in 0..MAX_VARINT_BYTES {
            self.check_available(1)?;
            let b = self.buffer[self.pos];
            self.pos += 1;

            // At the 10th byte (index 9), we've consumed 63 bits.
            // The 10th byte can only contribute 1 more bit (bit 63 of uint64).
            if i == 9 {
                // If continuation bit is set, we'd need 11+ bytes
                if b >= 0x80 {
                    return Err(Error::VarintOverflow);
                }
                // If data portion is > 1, value would overflow uint64
                if b > 1 {
                    return Err(Error::VarintOverflow);
                }
            }

            result |= ((b & 0x7f) as u64) << shift;
            if b & 0x80 == 0 {
                return Ok(result);
            }
            shift += 7;
        }

        Err(Error::VarintOverflow)
    }

    /// Reads a signed varint using ZigZag decoding.
    pub fn read_svarint(&mut self) -> Result<i32> {
        Ok(zigzag_decode_32(self.read_varint()?))
    }

    /// Reads a signed 64-bit varint using ZigZag decoding.
    pub fn read_svarint64(&mut self) -> Result<i64> {
        Ok(zigzag_decode_64(self.read_varint64()?))
    }

    /// Reads a V2 compact field tag.
    /// Returns a FieldTag with field_number=0 for end marker.
    pub fn read_tag(&mut self) -> Result<FieldTag> {
        let remaining = &self.buffer[self.pos..];
        if remaining.is_empty() {
            return Err(Error::buffer_underflow(1, 0));
        }

        let result = decode_compact_tag(remaining)
            .ok_or_else(|| Error::InvalidWireType(remaining[0]))?;

        self.pos += result.bytes_read;
        Ok(FieldTag::new(result.field_number, result.wire_type))
    }

    /// Checks if the next byte is the end marker without consuming it.
    pub fn peek_end_marker(&self) -> bool {
        self.pos < self.buffer.len() && self.buffer[self.pos] == END_MARKER
    }

    /// Checks if the given tag is an end marker.
    pub fn is_end_marker(tag: &FieldTag) -> bool {
        tag.field_number == 0
    }

    /// Reads a boolean.
    pub fn read_bool(&mut self) -> Result<bool> {
        Ok(self.read_byte()? != 0)
    }

    /// Reads a 32-bit signed integer.
    pub fn read_int32(&mut self) -> Result<i32> {
        self.read_svarint()
    }

    /// Reads a 64-bit signed integer.
    pub fn read_int64(&mut self) -> Result<i64> {
        self.read_svarint64()
    }

    /// Reads a 32-bit unsigned integer.
    pub fn read_uint32(&mut self) -> Result<u32> {
        self.read_varint()
    }

    /// Reads a 64-bit unsigned integer.
    pub fn read_uint64(&mut self) -> Result<u64> {
        self.read_varint64()
    }

    /// Reads a 32-bit float (IEEE 754, little-endian).
    pub fn read_float32(&mut self) -> Result<f32> {
        let bytes = self.read_bytes(4)?;
        Ok(f32::from_le_bytes([bytes[0], bytes[1], bytes[2], bytes[3]]))
    }

    /// Reads a 64-bit float (IEEE 754, little-endian).
    pub fn read_float64(&mut self) -> Result<f64> {
        let bytes = self.read_bytes(8)?;
        Ok(f64::from_le_bytes([
            bytes[0], bytes[1], bytes[2], bytes[3], bytes[4], bytes[5], bytes[6], bytes[7],
        ]))
    }

    /// Reads a fixed 32-bit value (little-endian).
    pub fn read_fixed32(&mut self) -> Result<u32> {
        let bytes = self.read_bytes(4)?;
        Ok(u32::from_le_bytes([bytes[0], bytes[1], bytes[2], bytes[3]]))
    }

    /// Reads a fixed 64-bit value (little-endian).
    pub fn read_fixed64(&mut self) -> Result<u64> {
        let bytes = self.read_bytes(8)?;
        Ok(u64::from_le_bytes([
            bytes[0], bytes[1], bytes[2], bytes[3], bytes[4], bytes[5], bytes[6], bytes[7],
        ]))
    }

    /// Reads a length-prefixed string.
    pub fn read_string(&mut self) -> Result<&'a str> {
        let length = self.read_varint()? as usize;
        let bytes = self.read_bytes(length)?;
        std::str::from_utf8(bytes).map_err(|_| Error::InvalidUtf8)
    }

    /// Reads length-prefixed bytes.
    pub fn read_length_prefixed_bytes(&mut self) -> Result<&'a [u8]> {
        let length = self.read_varint()? as usize;
        self.read_bytes(length)
    }

    /// Skips a field based on its wire type.
    pub fn skip_field(&mut self, wire_type: WireType) -> Result<()> {
        match wire_type {
            WireType::Varint | WireType::SVarint => {
                self.read_varint64()?; // Use 64-bit to handle large varints
            }
            WireType::Fixed64 => {
                self.check_available(8)?;
                self.pos += 8;
            }
            WireType::Bytes => {
                let length = self.read_varint()? as usize;
                self.check_available(length)?;
                self.pos += length;
            }
            WireType::Fixed32 => {
                self.check_available(4)?;
                self.pos += 4;
            }
        }
        Ok(())
    }

    /// Creates a sub-reader for reading nested messages.
    pub fn sub_reader(&mut self, length: usize) -> Result<Reader<'a>> {
        self.check_available(length)?;
        let sub = Reader::new(&self.buffer[self.pos..self.pos + length]);
        self.pos += length;
        Ok(sub)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_read_varint() {
        let mut reader = Reader::new(&[0]);
        assert_eq!(reader.read_varint().unwrap(), 0);

        let mut reader = Reader::new(&[1]);
        assert_eq!(reader.read_varint().unwrap(), 1);

        let mut reader = Reader::new(&[127]);
        assert_eq!(reader.read_varint().unwrap(), 127);

        let mut reader = Reader::new(&[0x80, 0x01]);
        assert_eq!(reader.read_varint().unwrap(), 128);

        let mut reader = Reader::new(&[0xac, 0x02]);
        assert_eq!(reader.read_varint().unwrap(), 300);
    }

    #[test]
    fn test_read_svarint() {
        let mut reader = Reader::new(&[0]);
        assert_eq!(reader.read_svarint().unwrap(), 0);

        let mut reader = Reader::new(&[1]);
        assert_eq!(reader.read_svarint().unwrap(), -1);

        let mut reader = Reader::new(&[2]);
        assert_eq!(reader.read_svarint().unwrap(), 1);
    }

    #[test]
    fn test_read_string() {
        let mut reader = Reader::new(&[5, b'h', b'e', b'l', b'l', b'o']);
        assert_eq!(reader.read_string().unwrap(), "hello");
    }

    #[test]
    fn test_read_v2_compact_tag() {
        // Field 1, wire type Varint: (1 << 4) | (0 << 1) = 0x10
        let mut reader = Reader::new(&[0x10]);
        let tag = reader.read_tag().unwrap();
        assert_eq!(tag.field_number, 1);
        assert_eq!(tag.wire_type, WireType::Varint);

        // Field 2, wire type SVarint: (2 << 4) | (4 << 1) = 0x28
        let mut reader = Reader::new(&[0x28]);
        let tag = reader.read_tag().unwrap();
        assert_eq!(tag.field_number, 2);
        assert_eq!(tag.wire_type, WireType::SVarint);

        // Field 15, wire type Bytes: (15 << 4) | (2 << 1) = 0xf4
        let mut reader = Reader::new(&[0xf4]);
        let tag = reader.read_tag().unwrap();
        assert_eq!(tag.field_number, 15);
        assert_eq!(tag.wire_type, WireType::Bytes);
    }

    #[test]
    fn test_read_v2_extended_tag() {
        // Field 16, wire type Varint: extended marker + varint 16
        // Extended: (0 << 1) | 1 = 0x01, then varint 16
        let mut reader = Reader::new(&[0x01, 16]);
        let tag = reader.read_tag().unwrap();
        assert_eq!(tag.field_number, 16);
        assert_eq!(tag.wire_type, WireType::Varint);

        // Field 1000, wire type SVarint
        // Extended: (4 << 1) | 1 = 0x09, then varint 1000 (0xe8, 0x07)
        let mut reader = Reader::new(&[0x09, 0xe8, 0x07]);
        let tag = reader.read_tag().unwrap();
        assert_eq!(tag.field_number, 1000);
        assert_eq!(tag.wire_type, WireType::SVarint);
    }

    #[test]
    fn test_read_end_marker() {
        let mut reader = Reader::new(&[END_MARKER]);
        let tag = reader.read_tag().unwrap();
        assert_eq!(tag.field_number, 0);
        assert!(Reader::is_end_marker(&tag));
    }

    #[test]
    fn test_read_write_roundtrip() {
        use crate::writer::Writer;

        let mut writer = Writer::new();
        writer.write_int32_field(1, -42).unwrap();
        writer.write_string_field(2, "hello").unwrap();
        writer.write_bool_field(3, true).unwrap();
        writer.write_end_marker().unwrap();

        let data = writer.into_bytes();
        let mut reader = Reader::new(&data);

        let tag1 = reader.read_tag().unwrap();
        assert_eq!(tag1.field_number, 1);
        assert_eq!(tag1.wire_type, WireType::SVarint);
        assert_eq!(reader.read_int32().unwrap(), -42);

        let tag2 = reader.read_tag().unwrap();
        assert_eq!(tag2.field_number, 2);
        assert_eq!(tag2.wire_type, WireType::Bytes);
        assert_eq!(reader.read_string().unwrap(), "hello");

        let tag3 = reader.read_tag().unwrap();
        assert_eq!(tag3.field_number, 3);
        assert_eq!(tag3.wire_type, WireType::Varint);
        assert!(reader.read_bool().unwrap());

        // Read end marker
        let end_tag = reader.read_tag().unwrap();
        assert!(Reader::is_end_marker(&end_tag));

        assert!(!reader.has_more());
    }

    #[test]
    fn test_peek_end_marker() {
        let mut reader = Reader::new(&[0x10, END_MARKER]);

        // First, peek - should not be end marker
        assert!(!reader.peek_end_marker());

        // Read the first tag
        let tag = reader.read_tag().unwrap();
        assert_eq!(tag.field_number, 1);

        // Now peek - should be end marker
        assert!(reader.peek_end_marker());

        // Read it
        let end_tag = reader.read_tag().unwrap();
        assert!(Reader::is_end_marker(&end_tag));
    }
}
